zaza/openstack/charm_tests/ceph/rbd_mirror/tests.py

```diff
@@ -124,6 +136,119 @@ def wait_for_mirror_state(self, state, application_name=None,
+                CIRROS_IMAGE_NAME,
+                LTS_IMAGE_NAME))
+            return openstack.get_images_by_name(glance, LTS_IMAGE_NAME)[0]
+
+        def create_volume_type(cinder):
+            try:
+                vol_type = cinder.volume_types.find(name='repl')
+            except cinder_exceptions.NotFound:
+                vol_type = cinder.volume_types.create('repl')
+                vol_type.set_keys(metadata={
+                    'volume_backend_name': 'cinder-ceph',
+                    'replication_enabled': '<is> True',
+                })
+            return vol_type
+
+        # NOTE(fnordahl): for some reason create volume from image often fails
+        # when run just after deployment is finished.  We should figure out
+        # why, resolve the underlying issue and then remove this.
+        #
+        # We do not use tenacity here as it will interfere with tenacity used
+        # in ``resource_reaches_status``
+        def create_volume(cinder, volume_params, retry=20):
+            if retry < 1:
+                return
+            volume = cinder.volumes.create(**volume_params)
+            try:
+                # Note(coreycb): stop_after_attempt is increased because using
+                # juju storage for ceph-osd backed by cinder on undercloud
+                # takes longer than the prior method of directory-backed OSD
+                # devices.
+                openstack.resource_reaches_status(
+                    cinder.volumes, volume.id, msg='volume',
+                    stop_after_attempt=20)
+                return volume
+            except AssertionError:
+                logging.info('retrying')
+                volume.delete()
+                return create_volume(cinder, volume_params, retry=retry - 1)
+
+        volume_params = {
+            'size': 8,
+            'name': 'zaza',
+        }
+        if from_image:
+            volume_params['imageRef'] = get_glance_image(session).id
+        cinder = openstack.get_cinder_session_client(session)
+        if self.get_cinder_rbd_mirroring_mode() == 'image':
+            volume_params['volume_type'] = create_volume_type(cinder).id
+
+        return create_volume(cinder, volume_params)
+
+    def failover_cinder_volume_host(self, cinder_client,
+                                    backend_name='cinder-ceph',
+                                    target_backend_id='ceph',
+                                    target_status='disabled',
+                                    target_replication_status='failed-over',
+                                    timeout=300):
+        """Failover Cinder volume host."""
+        host = 'cinder@{}'.format(backend_name)
+        logging.info(
+            'Failover Cinder host %s to backend_id %s',
+            host, target_backend_id)
+        cinder_client.services.failover_host(
+            host=host,
+            backend_id=target_backend_id)
+        start = time.time()
```

Comment:
I think we should use tenacity.Retry() here, except if there is a reason not to (like on line 184).

Reply:
I'll look to use the `tenacity` helper.